src/epanet.c

```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
+  
+  if(nodeType == EN_JUNCTION) {
+    Njuncs++;
+    n = Njuncs;
+    
+    struct Sdemand *demand;
+    demand = (struct Sdemand *) malloc(sizeof(struct Sdemand));
+    demand->Base = 5.0;
+    demand->Pat = 0;
+    demand->next = NULL;
+    Node[n].D = demand;
+    
+    int index;
+    // shift rest of Node array
+    for(index = Nnodes; index >= Njuncs; index--) {
+      ENHashTableUpdate(NodeHashTable, Node[index].ID, index+1);        
+      Node[index+1] = Node[index];
+      Coord[index+1] = Coord[index];
+    }
+    // shift indices of Tank array
+    for(index = 1; index <= Ntanks; index++) {
+      Tank[index].Node += 1;
+    }
+    
+    // shift indices of Pipes, if necessary
+    for(index = 1; index <= Npipes; index++) {
+      if(Link[index].N1 > Njuncs - 1) {
+        Link[index].N1 += 1;
+      }
+      if(Link[index].N2 > Njuncs - 1) {
+        Link[index].N2 += 1;
+      }
+    }
+  } else {
+    n = Nnodes+1;
+    Ntanks++;
+    
+  /* resize tanks array */
+    Tank = (Stank *)realloc(Tank, (Ntanks+1)*sizeof(Stank));
+    
+  /* set default values for new tank or reservoir */
+    Tank[Ntanks].Node = n;
+    Tank[Ntanks].Pat = 0;
+    if(nodeType == EN_TANK) {
+      Tank[Ntanks].A  = 1.0;
+    } else {
+      Tank[Ntanks].A  = 0;
+    }
+    Tank[Ntanks].Hmin = 0;
+    Tank[Ntanks].Hmax = 0;
+    Tank[Ntanks].H0 = 0;
+    Tank[Ntanks].Vmin = 0;
+    Tank[Ntanks].Vmax = 0;
+    Tank[Ntanks].V0 = 0;
+    Tank[Ntanks].Kb = 0;
+    Tank[Ntanks].V = 0;
+    Tank[Ntanks].C = 0;
+    Tank[Ntanks].Pat = 0;
+    Tank[Ntanks].Vcurve = 0;
+    Tank[Ntanks].MixModel = 0;
+    Tank[Ntanks].V1max = 10000;
+  }
+  
+  Nnodes++;
+
+  /* set default values for new node */
+  strncpy(Node[n].ID, id, MAXID);
+  
+  
+  Node[n].El = 0;
+  Node[n].S = NULL;
+  Node[n].C0 = 0;
+  Node[n].Ke = 0;
+  Node[n].Rpt = 0;
+  Coord[n].HaveCoords = FALSE;
+  Coord[n].X = 0;
+  Coord[n].Y = 0;
+  
+  /* Insert new node into hash table */
+  ENHashTableInsert(NodeHashTable, Node[n].ID, n);        /* see HASH.C */
+  return(0);
+}
+
+int DLLEXPORT ENaddlink(char *id, EN_LinkType linkType, char *fromNode, char *toNode)
+{
+  int i, n;
+  
+/* Check if a link with same id already exists */
+  if ( !Openflag ) return(102);
+  if ( ENgetlinkindex(id, &i) == 0 ) return(215);
+  
+/* Lookup the from and to nodes */
+  int N1 = ENHashTableFind(NodeHashTable, fromNode);
+  int N2 = ENHashTableFind(NodeHashTable, toNode);
+  
+  if (N1 == 0 || N2 == 0) {
+    return(203);
+  }
+  
+/* Check that id name is not too long */
+  if (strlen(id) > MAXID) return(250);
+  
+  Nlinks++;
+  n = Nlinks;
+  
+/* Grow arrays to accomodate the new value */
+  Link = (Slink *)realloc(Link, (Nlinks+1)*sizeof(Slink));
+  Q = (double *)realloc(Q, (Nlinks+1)*sizeof(double));
+  LinkSetting = (double *)realloc(LinkSetting, (Nlinks+1)*sizeof(double));
+  LinkStatus = (char *)realloc(LinkStatus, (Nlinks+1)*sizeof(char));
+  
+  strncpy(Link[n].ID, id, MAXID);

```

Comment:
Edge case here: if you specify a link with an ID of more than `MAXID` (31) characters, then the name will be truncated and you won't be able to access the new link by the ID you specified. How to handle?